### PR TITLE
chore: enable yamllint

### PR DESCRIPTION
**Pull request by ssbarnea**
This enables yamllint and ensures that indentation is consistent
across our files, fixing problem where prettier would reformat the
file on save.

### CODE REVIEW

**Review by webknjaz — CHANGES_REQUESTED** [review]
Please, don't add unnecessary indents to the sequences.

File: .yamllint [hunk 1]
```diff
@@ -0,0 +1,7 @@
+rules:
+  document-start: disable
+  indentation:
+    level: error
+    indent-sequences: true
```
Comment (webknjaz) [inline .yamllint:5]: This must be
```suggestion
    indent-sequences: false
```

**Comment by ssbarnea** [comment]
@webknjaz That is only imposing Ansible own standards, which is to indent sequences. In fact these files **already contained inconsistent sequences inside the same file**.

Look at PR template where labels sequence was not indented but body sequence was. Similar inside ci.yml where there was fix of the two. Basically you cannot say that is switching an indentation type with another, when in fact is consolidating it.

Before this change, editing a file in vscode would have introduced undesired indentation. At least @priyamsahoo was affected by this. In fact there are two extensions that can reformat `yaml` files when in vscode and both of them are behaving the same way (vscode-yaml and prettier).

I had a longer talk with @cidrblock today regarding which path we should take on this and some of the ideas that went out were:
* having different indentation rules based on where the yaml files are located on disk would only confuse users and force us to add a lot of complexity to our checks, shortly does not worth our time
* having ability to auto-reformat files from incoming pull-request is a valuable as almost nobody enjoys reformatting code to match a linter expectation. Having prettier auto-fix that on file save (or pre-commit) saves a lot of human time.
* we are part of Ansible team, so diverging from Ansible standards would be... not wise

Obviously that we will wait for input from the entire team on this one.

**Comment by webknjaz** [comment]
I asked some time ago, among the core team members, and most preferred no indent. So no, there's no such thing as an Ansible standard.

**Comment by ssbarnea** [comment]
There are serious problems especially with blocks when not using indentation. See https://docs.ansible.com/ansible/latest/user_guide/playbooks_blocks.html

When using blocks lack of indentation can often cause keys like `when:` to be wrongly applied to the wrong level, instead of applying to a block to apply to the last task inside the block, or the opposite. That is why very long time ago the indentation went towards always indenting sequences.

The fact that documentation uses that is not a random mistake.

**Comment by webknjaz** [comment]
The documentation represents something that tech writers chose to use, that's it. There is no official guidelines from the engineering.

**Comment by cidrblock** [comment]
Although I tend to not indent when making a sequence, I find both approaches legible.  Why I do appreciate is tools that ensure consistency in a repo without any manual work or intervention.

I'm a 0 on indent vs. not

+1 for a precommit something that ensures consistency

**Review by priyamsahoo — APPROVED** [review]
I am always in favor of consistency, irrespective of my opinion on indentations.

**Comment by tomaciazek** [comment]
> Please, don't add unnecessary indents to the sequences.

Just because something is optional, does not mean that it's unnecessary. Would you write code in Java/C++/C#/JavaScript without indentation (just because curly braces are enough)? I think that indentation of sequences provides much better legibility, especially with more complex cases, such as Ansible blocks.

**Review by tomaciazek — APPROVED** [review]

**Review by ganeshrn — APPROVED** [review]

**Review by cidrblock — APPROVED** [review]

**Comment by webknjaz** [comment]
> Just because something is optional, does not mean that it's unnecessary. Would you write code in Java/C++/C#/JavaScript without indentation (just because curly braces are enough)?

When I see an indent, I expect it to be a mapping. Sequences don't need an extra bit of syntax (both indent and a dash) — there should be one distinction, not multiple.

**Comment by tomaciazek** [comment]
> > Just because something is optional, does not mean that it's unnecessary. Would you write code in Java/C++/C#/JavaScript without indentation (just because curly braces are enough)?
> 
> When I see an indent, I expect it to be a mapping. Sequences don't need an extra bit of syntax (both indent and a dash) — there should be one distinction, not multiple. Besides, too much indentation contributes to overly long lines.

Is `a` a mapping:
```yaml
- a:
  - b
  - c
```
? No. Can you skip the indent? No. Hence, you can't possibly always expect an indent to be a mapping😉

Possibility to skip indentation on *some* sequences is inconsistent in itself, to say the least. In my opinion, the fact that creators of YAML made it possible did more harm than good, causing debates such as this. Sometimes less choice is better :)

Also, the indent in case of sequences is not an extra bit of syntax - it's a bit of formatting.

**Comment by webknjaz** [comment]
> ? No. Can you skip the indent? No. Hence, you can't possibly always expect an indent to be a mappingwink

But the indent is counted since the position of `a:` so there's no indent :shrug:

**Comment by webknjaz** [comment]
> Sometimes less choice is better :)

Sounds like it's one step away from dictatorship. If the "no-choice" option is wrong and there's no way to fix it — is it good or bad? Let the history lessons be the answer.

**Comment by webknjaz** [comment]
> Possibility to skip indentation on _some_ sequences is inconsistent in itself, to say the least.

Syntax-wise, it's not skipping in some because of the place the indentation counter starts.

**Comment by tomaciazek** [comment]
> > Sometimes less choice is better :)
> 
> Sounds like it's one step away from dictatorship. If the "no-choice" option is wrong and there's no way to fix it — is it good or bad? Let the history lessons be the answer.

That got political and escalated pretty quickly :P I'm by no means in favor of dictatorship and let's drop it at that.

**Comment by tomaciazek** [comment]
> > Possibility to skip indentation on _some_ sequences is inconsistent in itself, to say the least.
> 
> Syntax-wise, it's not skipping in some because of the place the indentation counter starts.

It's right in the specification: https://yaml.org/spec/1.2.2/
Just under `Example 8.21. Block Scalar Nodes`:
```
Since people perceive the “-” indicator as indentation, nested block sequences may be indented by one less space to compensate, except, of course, if nested inside another block sequence ([BLOCK-OUT context] versus [BLOCK-IN context]).
```

**Comment by webknjaz** [comment]
> That got political and escalated pretty quickly :P I'm by no means in favor of dictatorship and let's drop it at that.

Blame the Godwin's Law :) The internet is filled with rubbish references of disasters of the past (it gets me every time I see people jokingly "implementing" the stalin sort, for example). Besides, everything in the world is political.
FWIW I stand by the metaphor because this is how such cases _feel_, even when applied to software — and for me, it's a way to express the degree of how strongly I feel about this (I didn't mean to imply that you may have been a dictatorship supporter, I'm sorry if this upset you).
It's one thing if people choose a tool that doesn't have options and is completely different if it's forced upon them.

As for the yamllint configuration ­— it supports an option of per-file consistency. I could live with Ansible files indented but it's hard for me to accept this in other places. One of the reasons is that some of the configs get synchronized from other repositories and having the same indentation across projects certainly helps minify the merge conflicts in diffs.

**Comment by tomaciazek** [comment]
> Blame the Godwin's Law :) The internet is filled with rubbish references of disasters of the past (it gets me every time I see people jokingly "implementing" the stalin sort, for example). Besides, everything in the world is political. FWIW I stand by the metaphor because this is how such cases _feel_, even when applied to software — and for me, it's a way to express the degree of how strongly I feel about this (I didn't mean to imply that you may have been a dictatorship supporter, I'm sorry if this upset you). It's one thing if people choose a tool that doesn't have options and is completely different if it's forced upon them.
> 
> As for the yamllint configuration ­— it supports an option of per-file consistency. I could live with Ansible files indented but it's hard for me to accept this in other places. One of the reasons is that some of the configs get synchronized from other repositories and having the same indentation across projects certainly helps minify the merge conflicts in diffs.

Don't worry, I didn't get upset, nor feel offended - I just didn't want this discussion to get derailed into a political debate.

As for those synchronized config files: which ones are these? How many are there?

**Comment by webknjaz** [comment]
> As for those synchronized config files: which ones are these? How many are there?

Mostly top-level configs and things under `.github/` (not only YAML) + other config-like things (e.g. the Sphinx's `conf.py` etc.). Unfortunately, this synchronization is sporadic and manual right now but I'm hoping to adopt a technique of doing a `git merge` similar to what https://github.com/jaraco/skeleton provides (Jason actually uses it on hundreds of the repos, pretty fascinating stuff), except I want it to be a bit more flexible/composable/distributed. That is when I finally get to do this (automating it on the scale is non-trivial).
There are maintenance techniques that allow reducing merge conflicts that would only benefit from similar formatting.

**Comment by tomaciazek** [comment]
> Mostly top-level configs and things under `.github/` (not only YAML) + other config-like things (e.g. the Sphinx's `conf.py` etc.). Unfortunately, this synchronization is sporadic and manual right now but I'm hoping to adopt a technique of doing a `git merge` similar to what https://github.com/jaraco/skeleton provides (Jason actually uses it on hundreds of the repos, pretty fascinating stuff), except I want it to be a bit more flexible/composable/distributed. That is when I finally get to do this (automating it on the scale is non-trivial). There are maintenance techniques that allow reducing merge conflicts that would only benefit from similar formatting.

Well, that doesn't sound like that many YAML files, and I suppose the synchronization process could take care of the indentation. As for the per-file consistency, I fear that it would not improve the consistency much. Even if some less or more strict guidelines are set, more often than not we'd have PRs with change requests just because of wrong indentation, and that's just counterproductive.

**Comment by tomaciazek** [comment]
Another reason for which I stopped using un-indented sequences is issues with automatic code-folding in VS Code:
image

**Comment by ganeshrn** [comment]
When it comes to indentation and code formatting there will always to conflicting opinions :-).
Considering the number of project maintainers we have in favour to have this PR in we do have a clear majority. I will go ahead and merge this PR. If in future we as a team collectively realise it is not the right way to go we can always revert the change.

**Pull request merged**